website/blog/2018-06-26-on-consuming-and-publishing-es2015+-packages.md

```diff
@@ -0,0 +1,274 @@
+
+After all, seeing `"stage-0"` says nothing. My hope is that in making the decision to use proposal plugins explicit, people will have to learn what non-standard syntax they are opting into. More intentionally, this should lead to a better understanding of not just Babel but of JavaScript as a language and its development instead of just its usage.
+
+### Publishing Non-standard Syntax
+
+As a library author, publishing non-standard syntax is setting our users up for possible inconsistencies, refactoring, and breakage of their projects. Because a TC39 proposal (even at Stage 3) has a possibility of changing, it means we will inevitability have to change the library code. A "new" proposal doesn't mean the idea is fixed or certain but rather that we collectively want to explore the solution space.
+
+At least if we ship the compiled version, it will still work, and the library maintainer can change the output so that it compiles into code that works the same as before. Shipping the uncompiled version means that anyone consuming a package needs to have a build step to use it and needs to have the same configuration of Babel as us. This is in the same bucket as using TS/JSX/Flow: we wouldn't expect consumers to configure the same compiler environment just because we used them.
+
+### Conflating JavaScript Modules and ES2015+
+
+When we write `import foo from "foo"` or `require("foo")` and `foo` doesn't have an `index.js`, it resolves to the `main` field in the `package.json` of the module.
+
+Some tools like Rollup/webpack also read from another field called `module` (previously `jsnext:main`). It uses this to instead resolve to the JS Module file.
+
+- An example with [`redux`](https://github.com/reactjs/redux)
+
+```js
+// redux package.json
+{
+  ...
+  "main": "lib/redux.js", // ES5 + Common JS
+  "module": "es/redux.mjs", // ES5 + JS Modules
+}
+```
+
+This was introduced so that users could consume JS Modules (ESM).
+
+However, the sole intention of this field is ESM, not anything else. The [Rollup docs](https://github.com/rollup/rollup/wiki/pkg.module#wait-it-just-means-import-and-export--not-other-future-javascript-features) specify that the `module` field makes it clear that it's not intended for future JavaScript syntax.
+
+Despite this warning, package authors invariably conflate the use of ES modules with the JavaScript language level they authored it in.
+
+As such, we may need another way to signal the language level.
+
+#### Non-scalable Solutions?
+
+A common suggestion is for libraries to start publishing ES2015 under another field like `es2015`, e.g. `"es2015": "es2015/redux.mjs"`.
+
+```js
+// @angular/core package.json
+{
+  "main": "./bundles/core.umd.js",
+  "module": "./fesm5/core.mjs",
+  "es2015": "./fesm2015/core.js",
+  "esm5": "./esm5/core.mjs",
+  "esm2015": "./esm2015/core.mjs",
+  "fesm5": "./fesm5/core.mjs",
+  "fesm2015": "./fesm2015/core.mjs",
+}
+```
```

Comment:
☝️ While I get the push by some for `.mjs` the `@angular/core` and `redux` packages doesn't actually use it.

Reply:
yeah that was my question here https://github.com/babel/website/pull/1705#discussion_r198125693

Reply:
You could solve it by removing the angular.js and redux references (saying instead: some projects do this, others do that)

Reply:
> Going to change it to reflect the actual package e348fc0 then?

Okay.